SFT_README.md

```diff
@@ -0,0 +1,227 @@
+* We need to not re-use the last token
+
+For example (Copied from [config_files/training/config_lorem_ipsum_sft.yaml](config_files/training/config_lorem_ipsum_sft.yaml)):
+```yaml
+collate_fn:  
+  component_key: collate_fn
+  variant_key: mask_loss_collator_wrapper
+  config:
+    wrapped_collate_fn:  
+      component_key: collate_fn
+      variant_key: gpt_2_llm_collator
+      config:
+        sample_key: ${settings.referencing_keys.sample_key}
+        target_key: ${settings.referencing_keys.target_key}
+    target_keys_to_mask:
+      - ${settings.referencing_keys.target_key}
+    loss_ignore_index: -100
+    mask_tokens:
+      b_include_to_loss_token: ^
+      e_include_to_loss_token: $
+    tokenizer:
+      instance_key: tokenizer
+      pass_type: BY_REFERENCE
+```
+
+with
+
+```yaml
+train_dataset:
+  component_key: dataset
+  variant_key: packed_mem_map_dataset_continuous
+  config:
+    raw_data_path: ./data/lorem_ipsum_sft_40e0699/lorem_ipsum_sft_converted.40e0699.pbin
+    sequence_length: ${settings.training.sequence_length}
+    sample_key:  ${settings.referencing_keys.sample_key}
+    reuse_last_target: true
+```
+
+and with
+
+```yaml
+tokenizer:
+  component_key: tokenizer
+  variant_key: pretrained_hf_tokenizer
+  config:
+    pretrained_model_name_or_path: data/tokenizer/hf_gpt2
+    padding: false
+    truncation: false
+    special_tokens:
+      additional_special_tokens: 
+        - "^"
+        - "$"
+```
+
+Finally, run the instruction-tuning with the `run` entry point:
+```bash
+torch.distributed.run --nnodes 1 --nproc_per_node 2 --rdzv-endpoint=0.0.0.0:29555 src/modalities/__main__.py run --config_file_path config_files/training/config_lorem_ipsum_sft.yaml
+```
+
+> Note, that it is advised to add a special token (which is already known as non-special token to the tokenizers' voabulary) to indicate the end of an assistant turn within the `b_include_to_loss_token` and `e_include_to_loss_token` in your chat template. Change your chat template accordingly and make sure to inlcude this token as special token in the tokenizer configuration for the pbin file creation step and model training!
+
+#### A Note on Tokanization in Huggingface
+The special tokens are added to a [Trie](https://en.wikipedia.org/wiki/Trie), so that longer special tokens are split first and then shorter special tokens.
+Example from the huggingface documentation:
+
+```python
+>>> trie = Trie()
+>>> trie.split("[CLS] This is a extra_id_100")
+["[CLS] This is a extra_id_100"]
+
+>>> trie.add("[CLS]")
+>>> trie.add("extra_id_1")
+>>> trie.add("extra_id_100")
+>>> trie.split("[CLS] This is a extra_id_100")
+["[CLS]", " This is a ", "extra_id_100"]
+```
+
+When we add a special token, which exists within the tokenizer voabulary already, HF only marks it as special token (adds it to the trie).
+This means, if the sequence we add as special token already exists in the vocab, there is no need to resize the embedding matrix!
```

Comment:
so basically we reflect the special token by at least on of the original tokens in the vocabulary? 

For example, assuming "spec_tok_1" is supposed to be a special token, it gets tokenized to e.g.,  `["spec", "_", "tok", "_", "1"]` ? Since it has been added to the Trie, we therefore still know that it is special token but in training the special token would still be represented by a 5 token sequence and therefore the embedding matrix can remain unchanged? 

What happens if the special token cannot be represented?

Reply:
Not quite. 
Given `spec_tok_1` as a special token, which is part of the trie and given the text to tokenize being `Hi there spec_tok_1!`:
First, the split method of the trie is used to extract all special tokens. In our case, the split would look like this: `["Hi there", "spec_tok_1", "!"]`, then all strings that are not part of the trie, i.e., the normal tokens, are tokenized leading to e.g. `["Hi", "there", "spec_tok_1", "!"]`